### PR TITLE
8315034 : File.mkdirs() occasionally fails to create folders on Windows shared folder

### PR DESCRIPTION
**Pull request by weibxiao**
File.mkdirs() occasionally fails to create folders on Windows shared folders. It turned out that Windows API FindFirstFileW created the error ERROR_NO_MORE_FILES. In some of the cases with a valid file path, this error still returns this error code, which supposedly should not.

Adding this error code into the method of lastErrorReportable in the native code will be handled by JDK.

To test the fix, it needs to run three Java processes to create the folders on a remote file server.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315034](https://bugs.openjdk.org/browse/JDK-8315034): File.mkdirs() occasionally fails to create folders on Windows shared folder (**Bug** - P4)

### Reviewers
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16502/head:pull/16502` \
`$ git checkout pull/16502`

Update a local copy of the PR: \
`$ git checkout pull/16502` \
`$ git pull https://git.openjdk.org/jdk.git pull/16502/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16502`

View PR using the GUI difftool: \
`$ git pr show -t 16502`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16502.diff">https://git.openjdk.org/jdk/pull/16502.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16502#issuecomment-1833870672)

### CODE REVIEW

File: test/jdk/java/io/File/WindowsFileCreation.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ *  @bug 8315034
+ *  @summary Verify that File.mkdirs() will not fail to create folders on Windows shared folder
+ *  Passing remote shared folder as parent folder with the prefix of the subfolder
+ *  (ex, \\\\192.168.1.110\\sharedfolder\\temp) as first parameter,
+ *  second parameter will be the subfolder to be previously created folder(ex, letter a). For one loop, it will
+ *  two folders.
+ *  With different the parameters in command line, the test should run successfully without failure. This test
+ *  needs to run three Java processes simultaneously to verify the fix.
+ *  This test requires to run manually and the parent folder as Windows shared folder should be existing already.
+ */
```
Comment (AlanBateman) [inline test/jdk/java/io/File/WindowsFileCreation.java:34]: There's no `@test` tag so this test won't run, not even as a manual test. I would be inclined to just drop this test and add the "noreg-hard" label to the issue.

File: src/java.base/windows/native/libjava/canonicalize_md.c [hunk 2]
```diff
@@ -138,7 +138,8 @@ lastErrorReportable()
         || (errval == ERROR_BAD_NET_NAME)
         || (errval == ERROR_ACCESS_DENIED)
         || (errval == ERROR_NETWORK_UNREACHABLE)
-        || (errval == ERROR_NETWORK_ACCESS_DENIED)) {
+        || (errval == ERROR_NETWORK_ACCESS_DENIED)
+        || (errval == ERROR_NO_MORE_FILES)) {
```
Comment (AlanBateman) [inline src/java.base/windows/native/libjava/canonicalize_md.c:142]: @bplb Is this the same thing as JDK-8234363? Does this also need to be handled in the FindFirstFile usage in WindowsDirectoryStream?

Comment (bplb) [reply src/java.base/windows/native/libjava/canonicalize_md.c:142]: It seems likely. I don't recall having been able to replicate these problems however.

Comment (weibxiao) [reply src/java.base/windows/native/libjava/canonicalize_md.c:142]: I used to duplicate it quite easily on my personal laptop with JDK11 (one laptop with Windows 11, the other with Windows 10). After my laptop system got updated, it was almost impossible to recreate it with my two laptops. To verify this bug, we have to configure exactly the same testing environment as the customer was having.

**Comment by omasseau** [comment]
Hello, Has this been backported to Java 8 ? I can't really find the info.

**Comment by jaikiran** [comment]
Hello Olivier, this has been backported to 11, 17 and 21 update releases but not to JDK 8. I don't closely follow the update releases, so I don't know if there's a reason for that.

**Comment by bplb** [comment]
> Hello, Has this been backported to Java 8 ? I can't really find the info.

According to the issue [JDK-8321863](https://bugs.openjdk.org/browse/JDK-8321863) it looks like it has been so backported.

**Comment by weibxiao** [comment]
It was not backport to jdk8u. https://bugs.openjdk.org/browse/JDK-8321863 is for 8u411-perf, which is not same as jdk8u release.

The bug originally was reported in jdk11, not in jdk8. That’s the reason I did not backport the change to jdk8u.

From: Brian Burkhalter ***@***.***>
Date: Thursday, July 18, 2024 at 2:41 PM
To: openjdk/jdk ***@***.***>
Cc: Weibing Xiao ***@***.***>, Mention ***@***.***>
Subject: [External] : Re: [openjdk/jdk] 8315034 : File.mkdirs() occasionally fails to create folders on Windows shared folder (PR #16502)

Hello, Has this been backported to Java 8 ? I can't really find the info.

According to the issue JDK-8321863<https://bugs.openjdk.org/browse/JDK-8321863> it looks like it has been so backported.

**Comment by bplb** [comment]
> It was not backport to jdk8u. https://bugs.openjdk.org/browse/JDK-8321863 is for 8u411-perf, which is not same as jdk8u release.

Thanks for the clarification.

**Pull request closed without merging**